File: Programmers/Lev2/Common_Multiple/Common_Multiple.cpp
#include <string>
#include <vector>
#include <math.h>
#include <iostream>
using namespace std;

int get_common_divisor(int num1, int num2){
    int small_num = num1;
    int big_num = num2;
    int result = 1;
    
    if(num1 > num2){
        small_num = num2;
        big_num = num1;
    }
    
    for(int i=1; i< int(pow(small_num, 0.5)+1);i++){
        int div = small_num / i;
        int mod = small_num % i;
        if((mod == 0) && (big_num % i == 0)){
            result = i;
        }
        if((small_num % div == 0) && (big_num % div == 0)){
            if(result < div){
                result = div;
                break;
            }
        }
    }
    return result;
}

int solution(vector<int> arr) {
    int answer = 0;
    int length = arr.size();
    
    answer = arr[0];
    
    for(int i=1;i<length;i++){
        int common_divisor = get_common_divisor(answer, arr[i]);
        int div1 = answer / common_divisor;
        int div2 = arr[i] / common_divisor;
        
        answer = common_divisor * (div1) * (div2);
    }
    return answer;
}

int main(){
	vector<int> my_input = {2,6,8,14}; // 168

	cout << solution(my_input) << "\n";

	return 0;
}

File: Programmers/Lev2/Common_Multiple/Common_Multiple.py
def get_common_divsor(num1, num2):
    result = 1
    t_min = num1
    t_max = num2

    if num1 > num2:
        t_min = num2
        t_max = num1

    for i in range(1, int(t_min ** 0.5) + 1):
        t_div = t_min // i
        if t_min % i == 0 and t_max % i == 0:
            if result < i:
                result = i
        if t_min % t_div == 0 and t_max % t_div == 0:
            if result < t_div:
                result = t_div
                break
    return result


def solution(arr):
    answer = 0

    while len(arr) > 1:
        num1 = arr.pop()
        num2 = arr.pop()
        max_divisor = get_common_divsor(num1, num2)
        t_num1 = num1 // max_divisor
        t_num2 = num2 // max_divisor
        answer = t_num1 * t_num2 * max_divisor
        arr.append(answer)

    answer = arr.pop()
    return answer

if __name__ == '__main__':
    t_input = [[2,6,8,14], [1,2,3]] # 168, 6
    for i in range(0, len(t_input)):
        print(solution(t_input[i]))
